src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -776,6 +776,25 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITCompilationStarted(
       rewriter_wrapper.LoadInt32(method_def_md_token);
       rewriter_wrapper.LoadInt64(reinterpret_cast<INT64>(module_version_id_ptr));
 
+      // after the call is made, unbox any valuetypes
+      mdToken typeToken;
+      if (method_replacement.wrapper_method.method_signature.RetTypeIsObject()
+          && ReturnTypeIsValueTypeOrGeneric(module_metadata->metadata_import,
+                              module_metadata->metadata_emit,
+                              target.id,
+                              target.signature,
+                              &typeToken)) {
+        if (debug_logging_enabled) {
```

Comment:
The change here is rather small, which is good. After checking the two conditions 1) that we implicitly boxed something with the `object` return type and 2) we can determine that the return value is a value type or a generic type variable, we add the `unbox.any` instruction.